File: shortestWordDistance.py

from typing import List

"""
	Notes: use of a pynthon's version of map<int, vector<int>>
"""

class Solution:
    def shortestDistance(self, words: List[str], word1: str, word2: str) -> int:
        myDict = {}
        word1List = []
        word2List = []
        
        minimumValue = 4294967296
        
        for i in range(0, len(words)):
   		    myDict.setdefault(words[i], []).append(i) 
   
        word1List = myDict[word1]
        word2List = myDict[word2]
        
        for i in range(0, len(word1List)):
            for j in range(0, len(word2List)):
                if (abs(word1List[i] - word2List[j]) < minimumValue):
                    minimumValue = abs(word1List[i] - word2List[j])
        
        return minimumValue


words = ["practice", "makes", "perfect", "coding", "makes"]
word1 = "coding"
word2 = "makes"

myObj = Solution()
answer = myObj.shortestDistance(words, word1, word2)

print(answer)

File: shortestWordDistance.cpp
#include <iostream>
using namespace std;

#include <vector>
#include <map>
#include <climits>

/*
	Intuition:  Create a map, key string and value vector of int, then add
		    all element indexes to the same key string.
		    Traverse through both these vector<int> of indexes looking for the shortest 
		    value.

	RunTime: O(N^2)
	Space Complexity: O(N)

*/


class Solution {
public:
    int shortestDistance(vector<string>& words, string word1, string word2) 
    {
        map<string, vector<int>> myMap;
    vector<int> word1Vect;
    vector<int> word2Vect;
    
    int result = INT_MAX;
    
    for (int i = 0; i < words.size(); i++)
    {
        myMap[words[i]].push_back(i);
    }
    
    word1Vect = myMap[word1];
    word2Vect = myMap[word2];
    
    for (int i = 0; i < word1Vect.size(); i++)
    {
        for (int j = 0; j < word2Vect.size(); j++)
        {
            if (abs(word1Vect[i] - word2Vect[j]) < result)
            {
                result = abs(word1Vect[i] - word2Vect[j]);
            }
        }
    }
    
    return result;

    }
};

int main()
{
	vector<string> words = {"practice", "makes", "perfect", "coding", "makes"};
    	string word1 = "coding";
    	string word2 = "practice";

	Solution myObj;
	int answer = myObj.shortestDistance(words, word1, word2);
	
    	cout << answer << endl;	

	return 0;

}
